File: src/database/models/technician.model.ts
import mongoose, {
  Schema,
  Document,
  Model,
} from "mongoose";

export interface ITechnician
  extends Document {
  technicianId: string;
  name: string;
  phone: string;

  email?: string;

  vehicleType: string;

  serviceTypes: string[];

  availability:
    | "AVAILABLE"
    | "BUSY"
    | "OFFLINE";

  currentJob?: string | null;

  rating?: number;

  location: {
    type: string;
    coordinates: [number, number];
    address: string;
  };

  createdAt: Date;
  updatedAt: Date;
}

const technicianSchema =
  new Schema<ITechnician>(
    {
      technicianId: {
        type: String,
        required: true,
        unique: true,
        index: true,
      },

      name: {
        type: String,
        required: true,
        trim: true,
      },

      phone: {
        type: String,
        required: true,
      },

      email: {
        type: String,
      },

      vehicleType: {
        type: String,
        required: true,
      },

      serviceTypes: [
        {
          type: String,
        },
      ],

      availability: {
        type: String,
        enum: [
          "AVAILABLE",
          "BUSY",
          "OFFLINE",
        ],
        default: "AVAILABLE",
        index: true,
      },

      currentJob: {
        type: String,
        default: null,
      },

      rating: {
        type: Number,
        default: 5,
      },

      location: {
  type: {
    type: String,
    enum: ["Point"],
    required: true,
    default: "Point",
  },

  coordinates: {
    type: [Number],
    required: true,
  },

  address: {
    type: String,
    required: true,
  },
},
    },
    {
      timestamps: true,
    }
  );

// Geo index
technicianSchema.index({
  location: "2dsphere",
});

const TechnicianModel: Model<ITechnician> =
  mongoose.models.Technician ||
  mongoose.model<ITechnician>(
    "Technician",
    technicianSchema
  );

export default TechnicianModel;

File: technician/lib/models/technician_model.dart
class TechnicianModel {
  final String id;
  final String technicianId;
  final String name;
  final String phone;
  final String availability;
  final String? currentJob;
  final double rating;
  final String serviceArea;
  final String vehicleType;

  TechnicianModel({
    required this.id,
    required this.technicianId,
    required this.name,
    required this.phone,
    required this.availability,
    this.currentJob,
    required this.rating,
    required this.serviceArea,
    required this.vehicleType,
  });

  factory TechnicianModel.fromJson(Map<String, dynamic> json) {
    return TechnicianModel(
      id: json['_id'] ?? '',
      technicianId: json['technicianId'] ?? '',
      name: json['name'] ?? '',
      phone: json['phone'] ?? '',
      availability: json['availability'] ?? 'offline',
      currentJob: json['currentJob'],
      rating: (json['rating'] ?? 5.0) is int 
          ? (json['rating'] as int).toDouble() 
          : (json['rating'] ?? 5.0) as double,
      serviceArea: json['serviceArea'] ?? '',
      vehicleType: json['vehicleType'] ?? '',
    );
  }

  Map<String, dynamic> toJson() {
    return {
      '_id': id,
      'technicianId': technicianId,
      'name': name,
      'phone': phone,
      'availability': availability,
      'currentJob': currentJob,
      'rating': rating,
      'serviceArea': serviceArea,
      'vehicleType': vehicleType,
    };
  }
}
